### PR TITLE
HTML Reporter: Add a stop button

### PR DESCRIPTION
**Pull request by raytiley**
This adds a stop button that will set the queue length to 0.
This will allow the current test to finish executing,
but then finish the test run.

Fixes gh-928

### CODE REVIEW

**Comment by raytiley** [comment]
@trentmwillis Let me know if this approach flies and I'll figure out how to add a test. /CC @tundal45 @rwjblue

File: reporter/html.js [hunk 1]
```diff
@@ -108,6 +108,15 @@ function id( name ) {
 	return document.getElementById && document.getElementById( name );
 }
 
+function stopTests() {
+  var stopButton = id( "qunit-stop-tests-button" );
+  if ( stopButton ) {
+    stopButton.innerHTML = "Stopping";
```
Comment (leobalter) [inline reporter/html.js]: Where does it says it stopped? This is probably the place for this.

File: reporter/html.js [hunk 2]
```diff
@@ -108,6 +108,15 @@ function id( name ) {
...
+  }
+  QUnit.config.queue.length = 0;
+  return false;
```
Comment (leobalter) [inline reporter/html.js]: Can you fix the indentation? It should use tabs instead of spaces.

**Comment by leobalter** [comment]
It also needs to remove the stop button when the tests are done.

**Comment by raytiley** [comment]
I'll fix the indentation, not sure why my editoring config didn't work. So is the expectation that we add / remove the button during a test run, or should it always be int he bar but enabled / disabled? Is the toolbar the correct place for the button?

File: reporter/html.js (hunk 1 above)
Comment (raytiley) [inline reporter/html.js]: It says `Stop` when the button is created and added to the toolbar, and then set again to `Stop` when the tests are done.

**Comment by leobalter** [comment]
> So is the expectation that we add / remove the button during a test run, or should it always be int he bar but enabled / disabled? Is the toolbar the correct place for the button?

Good questions.

I like add / remove. Even a disabled button might be a distraction for me when the tests are done, but I won't complain if enabled / disabled is preferred here.

I think @trentmwillis might give a better answer for where we should place it, I would say yes to have in the toolbar if it better blends with ember-qunit. Otherwise, the toolbar seems ok for it.

**Comment by platinumazure** [comment]
What happens in terms of remaining tests when the stop button is clicked? Do the remaining tests disappear, or are they "skipped" or something else?

Personally I'd want to make sure that the tests are still visible, so a user can know that the test was known by QUnit but wasn't executed.

Also, what happens in terms of logging callbacks? Can we ensure that the correct callbacks are still run (including end of module and end of test run)?

**Comment by trentmwillis** [comment]
For location/appearance, I think it would make the most sense to put this in the `#qunit-testresult` area. My reasoning for this is that we display the current status there (e.g., "Running: some test") so it would make sense to have the button show up there when tests are running and then remove it when the tests have stopped. We could potentially even turn it into a toggle, where "stop" becomes "start". Thoughts?

**Comment by raytiley** [comment]
@trentmwillis so looking into this putting it in `#qunit-testresult` is kinda problematic as that element gets `innerHTML` set on it for each test. Do you think going setting up / tearing down the event handler for each event is worth it... I'm open to making anything work, I just don't want to make a drastic change and get it rejected as too much.

Thanks for the guidance.

**Comment by rwjblue** [comment]
>  putting it in #qunit-testresult is kinda problematic as that element gets innerHTML set on it for each test

Could we change this around a tad and make `#qunit-testresult` be something like:

```html
<div id="qunit-testresult">
  <div id="qunit-testresult-display"></div>
  <div id="qunit-testresult-controls"></div>
</div>
```

The code that currently sets `innerHTML` on `#qunit-testresult` would set it on `#qunit-testresult-display`, and the button would be stable in `#qunit-testresult-controls`.  

RE: design of this (which we haven't really talked about yet), the `#qunit-testresult-display` could be setup to be ~ 90ish% width and the controls would be the rest (with a `float: right`).

@trentmwillis / @leobalter - How does that sound to y'all?

**Comment by trentmwillis** [comment]
@raytiley what @rwjblue suggests is more-or-less what I had in mind.

**Comment by leobalter** [comment]
>Could we change this around a tad and make #qunit-testresult be something like: ...

this is good, @rwjblue. I like it

@raytiley how does it sound to you? Can we pursue this structure?

**Comment by raytiley** [comment]
Hey Everyone... I will for sure take care of this... Just short on time this week due to a unexpected funeral :/  I should be able to handle it this weekend.

**Comment by leobalter** [comment]
I'm sorry, @raytiley. Take it easy.

**Comment by leobalter** [comment]
Hi @raytiley. I imagine a weekend is never really enough, but just for a heads up if you prefer we can help addressing the review points.

**Comment by trentmwillis** [comment]
@raytiley any chance you'll get to this soon? If not, mind if I pick it up?

**Comment by raytiley** [comment]
@trentmwillis et al. Super sorry for the delay on this. I did some more work on it tonight to bring it more inline with what people were asking for. I also did some additional work so it's super clear that the test run was aborted (see screenshot).

I took a quick look at existing tests, but wasn't exactly sure how to go about testing this.

qunit example 2016-11-27 22-29-44

File: reporter/html.js [hunk 3]
```diff
@@ -610,14 +637,38 @@ QUnit.done( function( details ) {
...
+		}
+	}
 
-	if ( banner ) {
+	if ( banner && ( !stopButton || stopButton.disabled === false ) ) {
 		banner.className = details.failed ? "qunit-fail" : "qunit-pass";
 	}
 
+	if ( stopButton ) {
+    stopButton.parentNode.removeChild( stopButton );
```
Comment (rwjblue) [inline reporter/html.js]: This needs to be two tabs (looks like spaces ATM)

File: reporter/html.js [hunk 4]
```diff
@@ -610,14 +637,38 @@ QUnit.done( function( details ) {
...
+  }
```
Comment (rwjblue) [inline reporter/html.js]: Should be one tab here (looks like 2 spaces ATM)

File: reporter/html.js [hunk 5]
```diff
@@ -599,6 +625,7 @@ QUnit.begin( function( details ) {
 QUnit.done( function( details ) {
 	var banner = id( "qunit-banner" ),
 		tests = id( "qunit-tests" ),
+    stopButton = id( "qunit-stop-tests-button" ),
```
Comment (rwjblue) [inline reporter/html.js]: spaces -> tab

**Comment by trentmwillis** [comment]
I'd like to make sure the messaging is consistent, if we're aborting tests, then the button should say "abort". To me, if it says "stop", then that means potential for there to be a "start" coming after it; it's a "recoverable" action if you will.

To that end, I would actually rather we just stop the tests and not update the remaining tests. My thinking is two-fold:

1. This makes it easier for us to introduce a "re-start" feature in the future without altering behavior.
2. If you abort early on in a large test suite (several thousand tests for example), then all the updates can be annoying.

**Comment by leobalter** [comment]
@raytiley do you mind replacing Stop with Abort as suggested by @trentmwillis?

I'm +1 to land this after this small part is fixed.

**Comment by raytiley** [comment]
@trentmwillis @leobalter.... I'll plan on wrapping this up tonight. Can we clarify.

  ## Option A
 - Stop button changes to abort. Everything else remains the same

 ## Option B
 - Stop button stays stopped
 - Unrun tests are NOT shown
 - The status area updates to stay that tests were stopped (not aborted)

**Comment by trentmwillis** [comment]
@raytiley sorry for the delay, let's just do Option A. If we want to add `Start/Stop` functionality in the future, we can do it independent of the `Abort` functionality.

**Comment by raytiley** [comment]
@trentmwillis no worries. Option A done.

**Comment by trentmwillis** [comment]
This seems good to me. The styling of the button could use a bit of work, but we can address that in a follow up.

@leobalter let me know if we're good to merge.

**Pull request merged**